src/ring_theory/localization.lean

```diff
@@ -1,676 +1,539 @@
-
-@[simp] lemma lift_of (h : ∀ s ∈ S, is_unit (f s)) (a : α) :
-  lift f h (of a) = f a := lift'_of _ _ _ _
-
-@[simp] lemma lift_coe (h : ∀ s ∈ S, is_unit (f s)) (a : α) :
-  lift f h a = f a := lift'_of _ _ _ _
-
-@[simp] lemma lift'_comp_of (g : S → units β) (hg : ∀ s, (g s : β) = f s) :
-  lift' f g hg ∘ of = f := funext $ λ a, lift'_of _ _ _ a
-
-@[simp] lemma lift_comp_of (h : ∀ s ∈ S, is_unit (f s)) :
-  lift f h ∘ of = f := lift'_comp_of _ _ _
-
-@[simp] lemma lift'_apply_coe (f : localization α S → β) [is_ring_hom f]
-  (g : S → units β) (hg : ∀ s, (g s : β) = f s) :
-  lift' (λ a : α, f a) g hg = f :=
-have g = (λ s, (units.map' f : units (localization α S) → units β) (to_units s)),
-  from funext $ λ x, units.ext $ (hg x).symm ▸ rfl,
-funext $ λ x, localization.induction_on x
-  (λ r s, by subst this; rw [lift'_mk, ← (units.map' f).map_inv, units.coe_map'];
-    simp [is_ring_hom.map_mul f])
-
-@[simp] lemma lift_apply_coe (f : localization α S → β) [is_ring_hom f] :
-  lift (λ a : α, f a)
-    (λ s hs, is_unit.map' f (is_unit_unit (to_units ⟨s, hs⟩))) = f :=
-by rw [lift, lift'_apply_coe]
-
-/-- Function extensionality for localisations:
- two functions are equal if they agree on elements that are coercions.-/
-protected lemma funext (f g : localization α S → β) [is_ring_hom f] [is_ring_hom g]
-  (h : ∀ a : α, f a = g a) : f = g :=
-begin
-  rw [← lift_apply_coe f, ← lift_apply_coe g],
-  congr' 1,
-  exact funext h
-end
+To apply a localization map `f` as a function, we use `f.to_map`, as coercions don't work well for
+this structure.
 
-variables {α S T}
+We define an instance coercing a localization map `f` to its codomain `S` so that the `R`-algebra
+instance on `S` can 'know' the map needed to induce the `R`-algebra structure.
 
-def map (hf : ∀ s ∈ S, f s ∈ T) : localization α S → localization β T :=
-lift' (of ∘ f) (to_units ∘ subtype.map f hf) (λ s, rfl)
+## Tags
+localization, ring localization, commutative ring localization, characteristic predicate,
+commutative ring, field of fractions
+-/
+variables {R : Type*} [comm_ring R] (M : submonoid R) (S : Type*) [comm_ring S]
+          {P : Type*} [comm_ring P]
 
-instance map.is_ring_hom (hf : ∀ s ∈ S, f s ∈ T) : is_ring_hom (map f hf) :=
-lift'.is_ring_hom _ _ _
+open function
 
-@[simp] lemma map_of (hf : ∀ s ∈ S, f s ∈ T) (a : α) :
-  map f hf (of a) = of (f a) := lift'_of _ _ _ _
+set_option old_structure_cmd true
 
-@[simp] lemma map_coe (hf : ∀ s ∈ S, f s ∈ T) (a : α) :
-  map f hf a = (f a) := lift'_of _ _ _ _
+/-- The type of ring homomorphisms satisfying the characteristic predicate: if `f : R →+* S`
+satisfies this predicate, then `S` is isomorphic to the localization of `R` at `M`.
+We later define an instance coercing a localization map `f` to its codomain `S` so
+that the `R`-algebra instance on `S` can 'know' the map needed to induce the `R`-algebra
+structure. -/
+@[nolint has_inhabited_instance] structure localization_map
```

Comment:
If `f : R →+* S` is a ring hom, what is the easiest way to upgrade it to a localisation map, in the middle of a proof?
I guess we can use `{ .. f }`, but maybe we should have a tiny definition around this for better readability:
```
def localization_map.of_ring_hom (f) (hf : characteristic predicate) : localization_map :=
{ fld_name := hf,
  .. f }
```

Reply:
Ooh, my bad - that would be a useful def. I'll add it now